common/src/main/java/io/netty/util/internal/InternalThreadLocalMap.java

```diff
@@ -331,13 +334,18 @@ public boolean setIndexedVariable(int index, Object value) {
     private void expandIndexedVariableTableAndSet(int index, Object value) {
         Object[] oldArray = indexedVariables;
         final int oldCapacity = oldArray.length;
-        int newCapacity = index;
-        newCapacity |= newCapacity >>>  1;
-        newCapacity |= newCapacity >>>  2;
-        newCapacity |= newCapacity >>>  4;
-        newCapacity |= newCapacity >>>  8;
-        newCapacity |= newCapacity >>> 16;
-        newCapacity ++;
+        int newCapacity;
+        if (index < ARRAY_LIST_CAPACITY_EXPAND_THRESHOLD) {
+            newCapacity = index;
+            newCapacity |= newCapacity >>>  1;
+            newCapacity |= newCapacity >>>  2;
+            newCapacity |= newCapacity >>>  4;
+            newCapacity |= newCapacity >>>  8;
+            newCapacity |= newCapacity >>> 16;
+            newCapacity ++;
+        } else {
+            newCapacity = Integer.MAX_VALUE;
```

Comment:
`Integer.MAX_VALUE` cannot be used as an array length, in many JVMs. Their max array lengths is some some constant close to but less than that. I've used `Integer.MAX_VALUE - 1000` in other projects. We should define a max capacity constant, and use it here and in `nextVariableIndex`.

Reply:
Yes, but the problem is how to decide the proper value of max array length, seems it's [VM-dependent](https://stackoverflow.com/questions/3038392/do-java-arrays-have-a-maximum-size).

Reply:
We can pick a conservative minimum that works for all targets. I think it's fine to waste a few slots on that. Trying to compute a value by inspecting the JVM would be too complicated, in my opinion.

Reply:
Maybe we can take Java's `ArrayList` as reference, it choose `Integer.MAX_VALUE - 8` as the max array size. (http://hg.openjdk.java.net/jdk7/jdk7/jdk/file/tip/src/share/classes/java/util/ArrayList.java#l190)

Reply:
Yeah, that sounds like a good choice.